core/domain/state_domain.py

```diff
@@ -916,8 +940,917 @@ def get_all_outcomes(self) -> List[Outcome]:
+                                    f'HasElementXAtPositionY rule above.'
+                                )
+                    # `IsEqualToOrdering` should always come before
+                    # `IsEqualToOrderingWithOneItemAtIncorrectPosition` when
+                    # they are off by one value.
+                    item_to_layer_idx = {}
+                    for layer_idx, layer in enumerate(rule_spec.inputs['x']):
+                        for item in layer:
+                            item_to_layer_idx[item] = layer_idx
+
+                    for ele in equal_ordering_one_at_incorec_posn:
+                        is_off_by_one_value = False
+                        for layer_idx, layer in enumerate(ele):
+                            for item in layer:
+                                if layer_idx != item_to_layer_idx[item]:
+                                    is_off_by_one_value = True
+                        if is_off_by_one_value:
+                            raise utils.ValidationError(
+                                f'Rule - {rule_spec_index} of answer '
+                                f'group {ans_group_index} will never '
+                                f'be match because it is made '
+                                f'redundant by the IsEqualToOrdering'
+                                f'WithOneItemAtIncorrectPosition '
+                                f'rule above.'
+                            )
+        choices = self.customization_args['choices'].value
+        if len(choices) < 2:
+            raise utils.ValidationError(
+                'Atleast 2 choices should be there '
+                'in DragAndDropSortInput interaction.'
+            )
+        # All inputs should be non-empty and unique.
+        seen_choices = []
+        for choice in choices:
+            if choice.html in ('<p></p>', ''):
+                raise utils.ValidationError(
+                    'There should not be any empty '
+                    'choices in DragAndDropSortInput interaction.'
+                )
+            if choice.html in seen_choices:
+                raise utils.ValidationError(
+                    'There should not be any duplicate '
+                    'choices in DragAndDropSortInput interaction.'
+                )
+            seen_choices.append(choice.html)
+
+    def _validate_text_input(self) -> None:
+        """Validates the TextInput interaction.
+
+        Raises:
+            ValidationError. Text input height is not >= 1 and <= 10.
+            ValidationError. Duplicate rules are present.
+            ValidationError. `Contains` rule comes before these rules (Contains,
+                StartsWith and Equals) where `Contains` rule string is a
+                substring of other rules string.
+            ValidationError. `StartsWith` rule comes before the `Equals` and
+                another `StartsWith` rule where the `StartsWith` rule string is
+                a prefix of other rules string.
```

Comment:
Here are 4 errors but in the function there are 7

Reply:
Updated